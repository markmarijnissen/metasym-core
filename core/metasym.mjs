import _ from "lodash";

/**
 * This file contains only calculations (logic), and is seperated from input/output.
 * 
 * Input: Data is provided by the etl.mjs module (extract-transform-load) which uses the iconomi.mjs API
 * Output: Rebalancing is done by the rebalancing.mjs module.
 * 
 * General flow of calculations:
 * 
 * 1. scoreStrategies: calculates the score for every strategy (and the score for every asset in that strategy)
 * 2. scoreAssets: pivots the scored strategies to create a list of assets with a score (that is the sum of strategies).
 * 3. metasym: Takes the top N (10) assets and calculates the weights based on the score.
 */

export const validateNumber = (name, n, min = 0, max = 1) => {
    if (!_.isNumber(n)) throw new Error(`${name} is not a number`);
    if (n < min) throw new Error(`${name} is less than ${min}`);
    if (n > max) throw new Error(`${name} is more than ${max}`);
}

export const validateBool = (config, name) => {
    if (typeof _.get(config, name) !== "boolean") throw new Error(`${name} should be true or false`);
}

export const ensureDefaultConfig = (config, strategies) => {
    // const multiplier = {};
    // const verified = {};
    // if (strategies) {
    //     Object.keys(strategies).forEach(ticker => {
    //         multiplier[ticker] = 1.0;
    //         verified[ticker] = null;
    //     })
    // }
    _.defaultsDeep(config, {
        weights: [0, 0, 4, 3, 2, 1], // How to weigh returns for [ DAY, WEEK, 1M, 3M, 6M, 1Y ]
        filters: {
            verified: true,     // strategy must be manually verified
            positive: false,    // strategy score > 0
            mature: false,       // strategy has a 3M return value (suggesting it is >3 M old)
            minRebalanceCount: 0, // minimum rebalance count
            minCopiers: 0,         // minimum copiers
            minAUM: 0,             // minimum assets under management,
        },
        diversify: {            // prevent one asset from becoming dominant
            maxWeight: 1.0,     // max weight per asset
            excluded: ["USDT", "DAI", "TUSD"]   // except these assets!
        },
        multiplier: {},         // map from { ticker: multiplier }, use to boost/reduce/ignore strategies
        verified: {},           // manually verify strategies to be eligble for METASYM
        minStrategies: 10,
        assetMultiplier: {
            BNB: 0              // assets (e.g. coins) can also be boosted/reduced/ignored 
        },
        metasymSize: 10         // how many assets should be in the METASYM strategy?
    });

    // Validation
    validateNumber('diversify.maxWeight', config.diversify.maxWeight, 0, 1);
    validateNumber('metasymSize', config.metasymSize, 0, 1000);
    validateNumber('minStrategies', config.minStrategies, 1, 1000);
    validateNumber('filters.minAUM', config.filters.minAUM, 0, 1000000000);
    validateNumber('filters.minRebalanceCount', config.filters.minRebalanceCount, 0, 1000); 
    config.weights.forEach((w, i) => validateNumber(`weights[${i}]`, w, 0, 100));
    _.map(config.multiplier, (n, ticker) => validateNumber(`multiplier[${ticker}]`, n, 0, 100));
    _.map(config.assetMultiplier, (n, ticker) => validateNumber(`assetMultiplier[${ticker}]`, n, 0, 100));
    validateBool(config, "filters.verified");
    validateBool(config, "filters.positive");
    validateBool(config, "filters.mature");

    return config;
}

// Score the strategies based on a weighted average of the returns.
// Also score the assets in the portfolio by distributing the strategy score according to the asset weight.
// Will use strategy and asset multipliers to subjectively boost/reduce/ignore certain strategies and assets.
export const scoreStrategies = (strategies, config) => {
    if (!strategies || Object.keys(strategies).length < 400) {
        throw new Error("missing strategies");
    }
    if (!config) {
        throw new Error("missing config");
    }
    const { mature, verified, positive, minRebalanceCount, minCopiers, minAUM } = config.filters;
    const sumWeights = _.sum(config.weights.map(x => Number(x)));
    const returnFields = ["DAY", "WEEK", "MONTH", "THREE_MONTH", "SIX_MONTH", "YEAR"];

    // Pass 1 - Filter strategies & Calculate scores
    const result = _(strategies)
        .map((s,ticker) => {
            s.ticker = ticker;
            return s;
        })
        .filter(s => {
            // Apply filters
            return (!mature || !!s.statistics.returns.THREE_MONTH) &&
                (!verified || config.verified[s.ticker] === true) &&
                (minAUM === 0 || s.price.aum >= minAUM) &&
                (minRebalanceCount === 0 || s.structure.monthlyRebalancedCount >= minRebalanceCount)
        })
        .map(s => {
            // Calculate a weighted average based on DAY / WEEK / 1M / 3M / 6M / 1Y returns.
            s.score = 0;
            returnFields.forEach((field, i) => {
                s.score += (s.statistics.returns[field] || 0) * Number(config.weights[i]);
            });
            const rawscore = s.score / sumWeights;
            s.rscore = rawscore;
            
            // Apply the multiplier to boost/reduce/ignore strategies
            const multiplier = _.isNumber(config.multiplier[s.ticker]) ? config.multiplier[s.ticker] : 1.0;
            s.score = rawscore * multiplier;

            // Distribute the strategy-score over the assets
            s.structure.values.forEach(c => {
                const cmultiplier = _.isNumber(config.assetMultiplier[c.ticker]) ? config.assetMultiplier[c.ticker] : 1.0;
                c.rscore = c.rebalancedWeight * s.rscore;   //rscore = raw score
                c.score = c.rebalancedWeight * s.score * cmultiplier; // score = multiplier score
            });
            return s;
        })
        .filter(s => {
            // Apply filters
            return (!positive || s.rscore > 0);
        })
        .sortBy("score")
        .reverse()
        .value();

    // Pass 2 - Normalized & Fraction Scores
    const min = _.minBy(result, "score")?.score || 0;
    const max = _.maxBy(result, "score")?.score || 0;
    const sumScore = _.sum(result.map(s => s.score - min));
    const numStrategies = _.filter(result, s => s.score !== 0).length;
    if (sumScore === 0 || numStrategies < config.minStrategies) {
        console.warn(`Ensure at least ${config.minStrategies} strategies have config.verified[TICKER] = true and a config.multiplier[ticker] > 0`);
        throw new Error(`Less than ${config.minStrategies} strategies have a score > 0 (${numStrategies})`);
    }

    _.forEach(result, s => {
        // Normalized score [0...1]
        s.nscore = Number((s.score - min) / (max - min) * 1)

        // Fraction score; the sum of selected strategies is 1 (can be used for visualisation purposes later)
        s.fscore = Number((s.score - min) / sumScore);

        // Apply normalized and fraction score to the assets.
        s.structure.values.forEach(c => {
            const cmultiplier = _.isNumber(config.assetMultiplier[c.ticker]) ? config.assetMultiplier[c.ticker] : 1.0;
            c.nscore = c.rebalancedWeight * s.nscore * cmultiplier;
            c.fscore = c.rebalancedWeight * s.fscore * cmultiplier;
        });

    });

    return result;
}

// Pivot the list of scored strategies to a list of assets.
//
// Every asset may be use in zero, one more strategies.
// The asset score is based on the sum of strategies.
//
// USD stable coins are collapsed into USDC to provide a more honest scoring (instead of 'duplicates')
export const scoreAssets = (scoredStrategies) => {
    return _(scoredStrategies)
        .map(s => s.structure.values)
        .flatten()
        .filter(c => !!c.ticker)
        .map(c => {
            c.metasymTicker = c.ticker;
            if (["USDT", "DAI", "TUSD"].indexOf(c.ticker) >= 0) {
                c.metasymTicker = "USDC";
            }
            return c;
        })
        .groupBy("metasymTicker")
        .mapValues(assets => _.sumBy(assets, "nscore"))
        .map((nscore, ticker) => ({ ticker, nscore }))
        .sortBy("nscore")
        .reverse()
        .value();
}

const round = (x,n=2) => parseFloat(x.toFixed(n))

// Take the top N (10) assets from the scored assets list and calculate the weights.
export const metasymFromScoredAssets = (scoredAssets, config) => {
    const assets = _.cloneDeep(scoredAssets.slice(0, config.metasymSize));
    if (assets.length === 0 || assets.length < config.metasymSize) {
        throw new Error(`Less than ${config.metasymSize || 1} assets`);
    }
    const total = _.sumBy(assets, "nscore");
    assets.forEach(c => c.weight = round(c.nscore / total));
    
    if (config.diversify.maxWeight < 1.0) {
        const maxWeight = config.diversify.maxWeight;
        const assetsTooHigh = assets.filter(c => c.weight > maxWeight && config.diversify.excluded.indexOf(c.metasymTicker) < 0);
        const assetsTooHighTotal = _.sumBy(assetsTooHigh, "nscore"); 

        const assetsOK = assets.filter(c => c.weight <= maxWeight);
        const nscorePerPercentagePoint = _.sumBy(assetsOK, "nscore") / (1 - (assetsTooHigh.length * maxWeight));
        const newAssetsTooHighTotal = (assetsTooHigh.length * maxWeight) * nscorePerPercentagePoint;

        const nscoreCorrectionMultiplier = newAssetsTooHighTotal / assetsTooHighTotal;

        assetsTooHigh.forEach(c => c.correctedNscore = c.nscore * nscoreCorrectionMultiplier);
        assetsOK.forEach(c => c.correctedNscore = c.nscore);
        const correctedTotal = _.sumBy(assets, "correctedNscore");
        assets.forEach(c => c.weight = round(c.correctedNscore / correctedTotal));
    } 
    
    const sum = _.sumBy(assets, "weight");
    assets[0].weight = round(assets[0].weight + (1 - sum)); // rounding error correction
    return assets;
}

// Convienence function to perform all calculation steps
export const metasymFromStrategies = (strategies, config) => {
    ensureDefaultConfig(config);
    const scoredStrategies = scoreStrategies(strategies, config);
    const scoredAssets = scoreAssets(scoredStrategies)
    return metasymFromScoredAssets(scoredAssets, config);
}