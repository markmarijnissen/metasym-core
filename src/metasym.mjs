import _ from "lodash";

/**
 * This file contains only calculations (logic), and is seperated from input/output.
 * 
 * Input: Data is provided by the etl.mjs module (extract-transform-load) which uses the iconomi.mjs API
 * Output: Rebalancing is done by the rebalancing.mjs module.
 * 
 * General flow of calculations:
 * 
 * 1. scoreStrategies: calculates the score for every strategy (and the score for every asset in that strategy)
 * 2. scoreAssets: pivots the scored strategies to create a list of assets with a score (that is the sum of strategies).
 * 3. metasym: Takes the top N (10) assets and calculates the weights based on the score.
 */

export const ensureDefaultConfig = config => {
    return _.defaultsDeep(config, {
        weights: [0, 0, 4, 3, 2, 1], // How to weigh returns for [ DAY, WEEK, 1M, 3M, 6M, 1Y ]
        filters: {
            positive: false,    // strategy score > 0
            mature: false       // strategy has a 3M return value (suggesting it is >3 M old)
        },
        multiplier: {},         // map from { ticker: multiplier }, use to boost/reduce/ignore strategies
        verified: {},           // manually verify strategies to be eligble for METASYM
        defaultMultiplier: 1,   // new strategies are multiplied by 0 as default (= ignore)
        minStrategies: 10,
        assetMultiplier: {
            BNB: 0              // assets (e.g. coins) can also be boosted/reduced/ignored 
        },
    });
}

// Score the strategies based on a weighted average of the returns.
// Also score the assets in the portfolio by distributing the strategy score according to the asset weight.
// Will use strategy and asset multipliers to subjectively boost/reduce/ignore certain strategies and assets.
export const scoreStrategies = (strategies, config) => {
    if (!strategies || Object.keys(strategies).length < 400) {
        throw new Error("missing strategies");
    }
    if (!config) {
        throw new Error("missing config");
    }
    const { mature, positive } = config.filters;
    const sumWeights = _.sum(config.weights.map(x => Number(x)));
    const returnFields = ["DAY", "WEEK", "MONTH", "THREE_MONTH", "SIX_MONTH", "YEAR"];

    // Pass 1 - Filter strategies & Calculate scores
    const result = _(strategies)
        // .filter((s,ticker) => config.verified[ticker] === true)
        .map((s, ticker) => {
            s.ticker = ticker;

            // Calculate a weighted average based on DAY / WEEK / 1M / 3M / 6M / 1Y returns.
            s.score = 0;
            returnFields.forEach((field, i) => {
                s.score += (s.statistics.returns[field] || 0) * Number(config.weights[i]);
            });
            const rawscore = s.score / sumWeights;
            s.rscore = rawscore;
            
            // Apply the multiplier to boost/reduce/ignore strategies
            const multiplier = _.isNumber(config.multiplier[s.ticker]) ? config.multiplier[s.ticker] : config.defaultMultiplier;
            const verified_multiplier = config.verified[s.ticker] === true ? 1.0 : 0.0;
            s.score = rawscore * multiplier * verified_multiplier;

            // Distribute the strategy-score over the assets
            s.structure.values.forEach(c => {
                const cmultiplier = _.isNumber(config.assetMultiplier[c.ticker]) ? config.assetMultiplier[c.ticker] :  1.0;
                c.rscore = c.rebalancedWeight * s.rscore;   //rscore = raw score
                c.score = c.rebalancedWeight * s.score * cmultiplier; // score = multiplier score
            });
            return s;
        })
        .filter(s => {
            // Apply filters
            return (!mature || !!s.statistics.returns.THREE_MONTH) &&
                (!positive || s.rscore > 0);
        })
        .sortBy("score")
        .reverse()
        .value();

    // Pass 2 - Normalized & Fraction Scores
    const min = _.minBy(result, "score")?.score || 0;
    const max = _.maxBy(result, "score")?.score || 0;
    const sumScore = _.sum(result.map(s => s.score - min));
    const numStrategies = _.filter(result, s => s.score !== 0).length;
    if (sumScore === 0 || numStrategies < config.minStrategies) {
        console.warn(`Ensure at least ${config.minStrategies} strategies have config.verified[TICKER] = true and a config.multiplier[ticker] > 0`);
        throw new Error(`Less than ${config.minStrategies} strategies have a score > 0 (${numStrategies})`);
    }

    _.forEach(result, s => {
        // Normalized score [0...1]
        s.nscore = Number((s.score - min) / (max - min) * 1)

        // Fraction score; the sum of selected strategies is 1 (can be used for visualisation purposes later)
        s.fscore = Number((s.score - min) / sumScore);

        // Apply normalized and fraction score to the assets.
        s.structure.values.forEach(c => {
            const cmultiplier = _.isNumber(config.assetMultiplier[c.ticker]) ? config.assetMultiplier[c.ticker] : 1.0;
            c.nscore = c.rebalancedWeight * s.nscore * cmultiplier;
            c.fscore = c.rebalancedWeight * s.fscore * cmultiplier;
        });

    });

    return result;
}

// Pivot the list of scored strategies to a list of assets.
//
// Every asset may be use in zero, one more strategies.
// The asset score is based on the sum of strategies.
//
// USD stable coins are collapsed into USDC to provide a more honest scoring (instead of 'duplicates')
export const scoreAssets = (scoredStrategies) => {
    return _(scoredStrategies)
        .map(s => s.structure.values)
        .flatten()
        .filter(c => !!c.ticker)
        .map(c => {
            c.metasymTicker = c.ticker;
            if (["USDT", "DAI", "TUSD"].indexOf(c.ticker) >= 0) {
                c.metasymTicker = "USDC";
            }
            return c;
        })
        .groupBy("metasymTicker")
        .mapValues(assets => _.sumBy(assets, "nscore"))
        .map((nscore, ticker) => ({ ticker, nscore }))
        .sortBy("nscore")
        .reverse()
        .value();
}

const round = (x,n=4) => Number(x.toFixed(n))

// Take the top N (10) assets from the scored assets list and calculate the weights.
export const metasymFromScoredAssets = (scoredAssets, n = 10) => {
    const assets = scoredAssets.slice(0, n);
    const total = _.sumBy(assets, "nscore");
    const result = _(assets)
        .map((c) => {
            c.weight = round(c.nscore / total);
            return c;
        })
        .value();
    const sum = _.sumBy(result, "weight");
    result[0].weight = result[0].weight + (1 - sum);
    return result;
}

// Convienence function to perform all calculation steps
export const metasymFromStrategies = (strategies, config) => {
    ensureDefaultConfig(config);
    const scoredStrategies = scoreStrategies(strategies, config);
    const scoredAssets = scoreAssets(scoredStrategies)
    return metasymFromScoredAssets(scoredAssets, config.metasymSize);
}