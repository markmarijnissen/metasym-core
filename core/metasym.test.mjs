import { ensureDefaultConfig, scoreStrategies, scoreAssets, metasymFromScoredAssets, metasymFromStrategies } from "./metasym.mjs";

describe("config defaults", () => {
    const config = {};
    ensureDefaultConfig(config);

    test("An empty object defaults to a full config", () => {
        expect(config).toMatchSnapshot();
    });

    // test("config containts filters for verified, positieve, mature, minRebalanceCount and minAUM", () => {
    //     const config = {};
    //     ensureDefaultConfig(config);
    //     expect(config.filters).toHaveProperty("verified");
    //     expect(config.filters).toHaveProperty("positive");
    //     expect(config.filters).toHaveProperty("mature");
    //     expect(config.filters).toHaveProperty("minRebalanceCount");
    //     expect(config.filters).toHaveProperty("minAUM");
    // });

    test("default ETL expiration is 45 minutes", () => {
        expect(config.etlExpiration).toBe(45);
    });

    test("default minimum number of strategies to include is 10", () => {
        expect(config.minStrategies).toBe(10);
    });

    test("default dynamic rebalancing is at most every hour, at least every week, and happens for every change (> 0% diff)", () => {
        expect(config.rebalance.min).toBe(60);
        expect(config.rebalance.max).toBe(7 * 24 * 60);
        expect(config.rebalance.threshold).toBe(0.0);
    });

    test("default diversification is disabled, because maxWeight is 100%", () => {
        expect(config.diversify.maxWeight).toBe(1.0);
    });

    test("default diversification will not happen for USDT, DAI, TUSD and USDC", () => {
        expect(config.diversify.excluded).toEqual(["USDT", "DAI", "TUSD", "USDC"]);
    });

    test("default weights are 0 (day), 0 (week), 4 (1M), 3 (3M), 2 (6M) and 1 (1Y)", () => {
        expect(config.weights).toEqual([0, 0, 4, 3, 2, 1]);
    });
});

describe("config validation", () => {
    test("diversify.maxWeight must be between 0 and 1", () => {
        expect(() => ensureDefaultConfig({ diversify: { maxWeight: 2 } })).toThrow();
        expect(() => ensureDefaultConfig({ diversify: { maxWeight: 0.5 } })).not.toThrow();
        expect(() => ensureDefaultConfig({ diversify: { maxWeight: -1 } })).toThrow();
    });

    test("Size of the METASYM strategy must be between 3 and 1000", () => {
        expect(() => ensureDefaultConfig({ metasymSize: 10000 })).toThrow();
        expect(() => ensureDefaultConfig({ metasymSize: 10 })).not.toThrow();
        expect(() => ensureDefaultConfig({ metasymSize: 0 })).toThrow();
    });

    test("Minimum number of strategies to include must be between 1 and 1000", () => {
        expect(() => ensureDefaultConfig({ minStrategies: 10000 })).toThrow();
        expect(() => ensureDefaultConfig({ minStrategies: 10 })).not.toThrow();
        expect(() => ensureDefaultConfig({ minStrategies: 0 })).toThrow();
    });

    test("the minAUM filter should be 0 or more", () => {
        expect(() => ensureDefaultConfig({ filters: { minAUM: -100 } })).toThrow();
    });

    test("Minimum number of strategies to include must be between 0 and 1000", () => {
        expect(() => ensureDefaultConfig({ filters: { minRebalanceCount: 10000 } })).toThrow();
        expect(() => ensureDefaultConfig({ filters: { minRebalanceCount: 10 } })).not.toThrow();
        expect(() => ensureDefaultConfig({ filters: { minRebalanceCount: -1 } })).toThrow();
    });

    test("Weights must be more than 0, and 6 in total", () => {
        expect(ensureDefaultConfig({ weights: [1] }).weights).toEqual([1, 0, 4, 3, 2, 1]);
        expect(() => ensureDefaultConfig({ weights: [0, 0, 0, 1, 0, 0, 0, 0] })).toThrow();
        expect(() => ensureDefaultConfig({ weights: [0, 0, 0, 0, 0, 0] })).toThrow();
        expect(() => ensureDefaultConfig({ weights: [0, 0, 0, -1, 0, 0] })).toThrow();
    });

    test("Verified filter should be true or false", () => {
        expect(() => ensureDefaultConfig({ filters: { verified: "true" } })).toThrow();
    });

    test("Positive filter should be true or false", () => {
        expect(() => ensureDefaultConfig({ filters: { positive: "true" } })).toThrow();
    });

    test("Mature filter should be true or false", () => {
        expect(() => ensureDefaultConfig({ filters: { mature: "true" } })).toThrow();
    });
});

describe("score strategies", () => {
    const fakeStrategy = (ticker) => {
        return {
            "manager": `${ticker} Manager`,
            "name": `${ticker} Strategy`,
            "price": {
                "aum": 143266.80495417636,
                "change12m": 0.2207,
                "change1m": -0.1956,
                "change24h": 0.0129,
                "change3m": -0.3401,
                "change6m": -0.2593,
                "change7d": -0.0769,
                "changeAll": 8.4665,
                "currency": "USD",
                "price": 9.49733363,
                "ticker": `${ticker}`
            },
            "statistics": {
                "currency": "USD",
                "maxDrawdown": {
                    "ALL_TIME": -0.6104,
                    "MONTH": -0.2469,
                    "SIX_MONTH": -0.4712,
                    "THREE_MONTH": -0.4162,
                    "WEEK": -0.1073,
                    "YEAR": -0.5825
                },
                "returns": {
                    "ALL_TIME": 18,
                    "DAY": 0,
                    "MONTH": 1,
                    "SIX_MONTH": 6,
                    "THREE_MONTH": 3,
                    "WEEK": 0.25,
                    "YEAR": 12
                },
                "ticker": `${ticker}`,
                "volatility": {
                    "ALL_TIME": 0.0489,
                    "MONTH": 0.0409,
                    "SIX_MONTH": 0.0392,
                    "THREE_MONTH": 0.0389,
                    "WEEK": 0.0609,
                    "YEAR": 0.0488
                }
            },
            "structure": {
                "lastRebalanced": 1646425528,
                "monthlyRebalancedCount": 7,
                "values": [{
                    "name": "Tether",
                    "profit": 9.276928885790846E-5,
                    "rebalancedWeight": 0.5,
                    "targetWeight": 0.4879,
                    "ticker": "USDT"
                }, {
                    "name": "PAX Gold",
                    "profit": 0.04686694162197974,
                    "rebalancedWeight": 0.5,
                    "targetWeight": 0.5121,
                    "ticker": "PAXG"
                }]
            }
        }

    }

    test("must include a minimum number of strategies", () => {
        expect(() => scoreStrategies(null, config)).toThrow();
        expect(() => scoreStrategies({ A: fakeStrategy("A") }, ensureDefaultConfig({ minStrategies: 2, verified: { A: true } }))).toThrow();
        expect(() => scoreStrategies(
            { A: fakeStrategy("A"), B: fakeStrategy("B") },
            ensureDefaultConfig({ minStrategies: 1, verified: { A: true, B: true } }))
        ).not.toThrow();
    });

    test("must include a config", () => {
        expect(() => scoreStrategies({ A: fakeStrategy("A"), B: fakeStrategy("B") }, null)).toThrow();
    });

    test.todo("calculate a raw score based on returns and weights");
    test.todo("use the multiplier to calculate final score based on raw score");
    test.todo("calculate a normalized score [0..1]");
    test.todo("calculate a fractional score, where sum of strategy-scores and sum of all asset-scores is always 1");
    test.todo("distribute the score over the assets, based on their weight")

    test.todo("only score verified strategies when verified filter is true");
    test.todo("score all strategies when verified filter is false");

    test.todo("only score positive strategies when positive filter is true");
    test.todo("score all strategies when positive filter is false");

    test.todo("score only strategies that have a rebalance count of equal or more then filters.minRebalanceCount");
    test.todo("score only strategies that have a AUM of equal or more then filters.minAUM");

    test.todo("only score strategies with 3M returns when mature filter is true");
    test.todo("score all strategies when mature filter is false");
});