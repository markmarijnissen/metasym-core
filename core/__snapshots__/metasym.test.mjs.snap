// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config defaults An empty object defaults to a full config 1`] = `
Object {
  "assetMultiplier": Object {},
  "diversify": Object {
    "excluded": Array [
      "USDT",
      "DAI",
      "TUSD",
      "USDC",
    ],
    "maxWeight": 1,
  },
  "etlExpiration": 45,
  "filters": Object {
    "mature": false,
    "minAUM": 0,
    "minCopiers": 0,
    "minRebalanceCount": 0,
    "positive": false,
    "verified": true,
  },
  "metasymSize": 10,
  "minStrategies": 10,
  "multiplier": Object {},
  "rebalance": Object {
    "max": 10080,
    "min": 60,
    "threshold": 0,
  },
  "verified": Object {},
  "weights": Array [
    0,
    0,
    4,
    3,
    2,
    1,
  ],
}
`;
